modules/channel-teams/src/backend/client.ts

```diff
@@ -0,0 +1,181 @@
+        payload: { text: ctx.activity.text },
+        preview: ctx.activity.text,
+        threadId,
+        target: ctx.activity.from.id,
+        ...args
+      })
+    )
+  }
+
+  public async sendOutgoingEvent(event: sdk.IO.Event): Promise<void> {
+    const messageType = event.type === 'default' ? 'text' : event.type
+
+    if (!_.includes(outgoingTypes, messageType)) {
+      throw new Error('Unsupported event type: ' + event.type)
+    }
+
+    const ref = this.conversationsRefs[event.threadId]
+
+    let msg: any = event.payload // TODO: place this logic in builtin with content-types
+    if (msg.type === 'typing') {
+      msg = {
+        type: 'typing'
+      }
+    } else if (msg.type === 'carousel') {
+      msg = this._sendCarousel(event)
+    } else if (msg.quick_replies && msg.quick_replies.length) {
+      msg = this._sendChoice(event)
+    }
+
+    await this.adapter.continueConversation(ref, async (turnContext: TurnContext) => {
+      try {
+        await turnContext.sendActivity(msg)
+      } catch (err) {
+        const message = `The following error occured when sending a payload of type ${msg.type} to MS Botframework: ${
+          err.message
+        }`
+        this.bp.logger.error(message, err)
+        throw err
+      }
+    })
+  }
+
+  private _sendChoice(event: sdk.IO.Event) {
+    return {
+      text: event.payload.text,
+      attachments: [
+        CardFactory.heroCard(
+          '',
+          CardFactory.images([]),
+          CardFactory.actions(
+            event.payload.quick_replies.map(reply => {
+              return {
+                title: reply.title,
+                type: 'messageBack',
+                value: reply.payload,
+                text: reply.payload,
+                displayText: reply.title
+              }
+            })
+          )
+        )
+      ]
+    }
+  }
+
+  private _sendCarousel(event: sdk.IO.Event) {
+    return {
+      type: 'message',
+      attachments: event.payload.elements.map(card => {
+        let contentUrl = card.picture
+
+        return CardFactory.heroCard(
+          card.title,
+          CardFactory.images([contentUrl]),
+          CardFactory.actions(
+            card.buttons.map(button => {
+              if (button.type === 'open_url') {
+                return {
+                  type: 'openUrl',
+                  value: button.url,
+                  title: button.title
+                }
+              } else if (button.type === 'say_something') {
+                return {
+                  type: 'messageBack',
+                  title: button.title,
+                  value: button.text,
+                  text: button.text,
+                  displayText: button.text
+                }
+              } else if (button.type === 'postback') {
+                return {
+                  type: 'messageBack',
+                  title: button.title,
+                  value: button.payload,
+                  text: button.payload
```

Comment:
Shouldn't text be the title of the button? I'm a bit puzzled by the differences between title, text, displayText, and

Reply:
I think value might be unused but not quite sure. Here's the signification of all fields:

title : Appears as the button label.

displayText : Optional. Echoed by the user into the chat stream when the action is performed. This text is not sent to your bot.

value : Sent to your bot when the action is performed. You can encode context for the action, such as unique identifiers or a JSON object.

text : Sent to your bot when the action is performed. Use this property to simplify bot development: Your code can check a single top-level property to dispatch bot logic.

https://docs.microsoft.com/en-us/microsoftteams/platform/concepts/cards/cards-actions